### PR TITLE
add KITTI detection and segmentation

### PR DESCRIPTION
**Pull request by zmaslova**
### Summary

Related #280 

### Checklist

- [X] I submit my changes into the `develop` branch
- [X] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [X] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [X] I have added tests to cover my changes
- [x] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [X] I submit _my code changes_ under the same [MIT License](
  https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [X] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: datumaro/plugins/kitti_format/converter.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+                    os.makedirs(osp.dirname(labels_file), exist_ok=True)
+                    with open(labels_file, 'w', encoding='utf-8') as f:
+                        for bbox in bboxes:
+                            label_line = [-1] * 15
+                            label_line[0] = self.get_label(bbox.label)
+                            truncated = KittiPath.DEFAULT_TRUNCATED
+                            occluded = KittiPath.DEFAULT_OCCLUDED
+                            if 'truncated' in bbox.attributes and \
+                                self._allow_attributes:
+                                truncated = float(bbox.attributes['truncated'])
```
Comment (zhiltsov-max) [inline datumaro/plugins/kitti_format/converter.py:140]:
```suggestion
                                truncated = cast(bbox.attributes.get('truncated'), float, KittiPath.DEFAULT_TRUNCATED)
```

File: datumaro/plugins/kitti_format/converter.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+                            if 'occluded' in bbox.attributes and \
+                                self._allow_attributes:
+                                occluded = int(bbox.attributes['occluded'])
+                            label_line[1] = truncated
+                            label_line[2] = occluded
+                            x, y, h, w = bbox.get_bbox()
+                            label_line[4:8] = x, y, x+h, y+w
```
Comment (zhiltsov-max) [inline datumaro/plugins/kitti_format/converter.py:147]:
```suggestion
                            label_line[4:8] = [x, y, x + h, y + w]
```

File: datumaro/plugins/kitti_format/converter.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+        elif isinstance(label_map_source, str) and osp.isfile(label_map_source):
+            label_map = parse_label_map(label_map_source)
+
+        else:
+            raise Exception("Wrong labelmap specified, "
+                "expected one of %s or a file path" % \
+                ', '.join(t.name for t in LabelmapType))
+
+        self._categories = make_kitti_categories(label_map)
```
Comment (zhiltsov-max) [inline datumaro/plugins/kitti_format/converter.py:199]: Does this format require black color to have index 0?

File: datumaro/plugins/kitti_format/extractor.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            instances_path = osp.join(self._path, KittiPath.INSTANCES_DIR,
+                image_name)
+            if self._task == KittiTask.segmentation and \
+                osp.isfile(instances_path):
+                instances_mask = load_image(instances_path, dtype=np.int32)
+                segm_ids = np.unique(instances_mask)
+                for segm_id in segm_ids:
+                    semantic_id = segm_id >> 8
+                    ann_id = segm_id % 256
+                    isCrowd = bool(ann_id == 0)
```
Comment (zhiltsov-max) [inline datumaro/plugins/kitti_format/extractor.py:71]:
```suggestion
                    isCrowd = (ann_id == 0)
```

File: datumaro/plugins/kitti_format/importer.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+        for task in KittiPath.TASK_DIR:
+            subset_paths = glob(osp.join(path, '**', KittiPath.TASK_DIR[task]),
+                recursive=True)
+            for subset_path in subset_paths:
+                path = osp.normpath(osp.join(subset_path, ".."))
+                subset_name = osp.splitext(osp.basename(path))[0]
+                subsets.setdefault(subset_name, {})[task] = path
+
+        return subsets
```
Comment (zhiltsov-max) [inline datumaro/plugins/kitti_format/importer.py:77]: Please add subformat importers as in #281

File: docs/formats/kitti_user_manual.md [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+```
+└─ Dataset/
+    ├── testing/
+    │   └── image_2/
+    │       ├── <name_1>.png
+    │       ├── <name_2>.png
+    │       └── ...
+    ├── training/
+    │   ├── image_2/ # left color camera images
+    │   │   ├── <name_1>.png
```
Comment (zhiltsov-max) [inline docs/formats/kitti_user_manual.md:59]: I suspect these images can have any extensions

File: docs/formats/kitti_user_manual.md [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+There are few examples of using Datumaro operations to solve
+particular problems with KITTI dataset:
+
+### Example 1. How to load an original KITTI dataset and convert to Cityscapes
+
+```bash
+datum create -o project
+datum add path -p project -f kitti ./KITTI/
+datum stats -p project
+datum export -p final_project -o dataset -f cityscapes --overwrite -- --save-images
```
Comment (zhiltsov-max) [inline docs/formats/kitti_user_manual.md:177]:
```suggestion
datum export -p final_project -o dataset -f cityscapes -- --save-images
```

File: tests/test_kitti_format.py [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+
+    def test_can_save_and_load_image_with_arbitrary_extension(self):
+        class TestExtractor(TestExtractorBase):
+            def __iter__(self):
+                return iter([
+                    DatasetItem(id='q/1', image=Image(path='q/1.JPEG',
+                        data=np.zeros((4, 3, 3)))),
+
+                    DatasetItem(id='a/b/c/2', image=Image(
+                             path='a/b/c/2.bmp', data=np.ones((1, 5, 3))
```
Comment (zhiltsov-max) [inline tests/test_kitti_format.py:420]: Check images are saved with original extensions

**Review by zhiltsov-max — CHANGES_REQUESTED** [review]

File: datumaro/plugins/kitti_format/converter.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+                        compiled_class_mask.class_mask, apply_colormap=False,
+                        dtype=np.int32)
+
+                    compiled_instance_mask = CompiledMask.from_instance_masks(masks,
+                        instance_labels=[(m.label << 8) + m.id for m in masks])
+                    inst_path = osp.join(subset_name,
+                        KittiPath.INSTANCES_DIR, image_name)
+                    self.save_mask(osp.join(self._save_dir, inst_path),
+                        compiled_instance_mask.class_mask, apply_colormap=False,
+                        dtype=np.int32)
```
[the remark this answers is not in the document]
Comment (zmaslova) [reply datumaro/plugins/kitti_format/converter.py:124]: kitti segmentation is quite the same as cityscapes (different folders layout, but same masks behavior), so it's correct and necessary

File: datumaro/plugins/kitti_format/converter.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+    def apply(self):
+        os.makedirs(self._save_dir, exist_ok=True)
+
+        for subset_name, subset in self._extractor.subsets().items():
+            if KittiTask.segmentation in self._tasks:
+                os.makedirs(osp.join(self._save_dir, subset_name,
+                    KittiPath.INSTANCES_DIR), exist_ok=True)
+            for item in subset:
+                image_name = item.id+KittiPath.IMAGE_EXT
```
[the remark this answers is not in the document]
Comment (zmaslova) [reply datumaro/plugins/kitti_format/converter.py]: Done

File: datumaro/plugins/kitti_format/converter.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+                    self._save_image(item, osp.join(self._save_dir, image_path))
+
+                masks = [a for a in item.annotations
+                    if a.type == AnnotationType.mask]
+                if masks and KittiTask.segmentation in self._tasks:
+                    compiled_class_mask = CompiledMask.from_instance_masks(masks,
+                        instance_labels=[self._label_id_mapping(m.label)
+                            for m in masks])
+                    color_mask_path = osp.join(subset_name,
+                        KittiPath.SEMANTIC_RGB_DIR, image_name)
```
[the remark this answers is not in the document]
Comment (zmaslova) [reply datumaro/plugins/kitti_format/converter.py]: Done

File: datumaro/plugins/kitti_format/converter.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+                    self.save_mask(osp.join(self._save_dir, color_mask_path),
+                        compiled_class_mask.class_mask)
+
+                    labelids_mask_path = osp.join(subset_name,
+                        KittiPath.SEMANTIC_DIR, image_name)
```
[the remark this answers is not in the document]
Comment (zmaslova) [reply datumaro/plugins/kitti_format/converter.py]: Done

File: docs/formats/kitti_user_manual.md (hunk 6 above)
Comment (zmaslova) [reply docs/formats/kitti_user_manual.md]: Updated

File: docs/formats/kitti_user_manual.md (hunk 7 above)
Comment (zmaslova) [reply docs/formats/kitti_user_manual.md]: Removed

File: tests/test_kitti_format.py (hunk 8 above)
Comment (zmaslova) [reply tests/test_kitti_format.py:447]: Done

File: datumaro/plugins/kitti_format/importer.py (hunk 5 above)
Comment (zmaslova) [reply datumaro/plugins/kitti_format/importer.py:77]: Added (also updated tests)

File: datumaro/plugins/kitti_format/importer.py [hunk 13]
```diff
@@ -75,3 +75,11 @@ def find_sources(path):
                 subsets.setdefault(subset_name, {})[task] = path
 
         return subsets
+
+class KittiDetectionImporter(KittiImporter):
+    _TASK = KittiTask.detection
+    _TASKS = { _TASK: KittiImporter._TASKS[_TASK] }
```
Comment (zhiltsov-max) [inline datumaro/plugins/kitti_format/importer.py:81]: This variable is not used in `find_sources()`

File: tests/test_kitti_format.py (hunk 8 above)
Comment (zhiltsov-max) [reply tests/test_kitti_format.py:447]: Maybe, its better to add an `osp.isfile()` check below?

Comment (zmaslova) [reply tests/test_kitti_format.py:447]: Added

File: datumaro/plugins/kitti_format/importer.py (hunk 13 above)
Comment (zmaslova) [reply datumaro/plugins/kitti_format/importer.py:80]: Updated

File: datumaro/plugins/kitti_format/converter.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+)
+
+LabelmapType = Enum('LabelmapType', ['kitti', 'source'])
+
+class KittiConverter(Converter):
+    DEFAULT_IMAGE_EXT = KittiPath.IMAGE_EXT
+
+    @staticmethod
+    def _split_tasks_string(s):
+        return [KittiTask[i.strip()] for i in s.split(',')]
```
[the remark this answers is not in the document]
Comment (zmaslova) [reply datumaro/plugins/kitti_format/converter.py]: Done

File: datumaro/plugins/kitti_format/importer.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+        from datumaro.components.project import Project # cyclic import
+        project = Project()
+
+        subsets = self.find_sources(path)
+
+        if len(subsets) == 0:
+            raise Exception("Failed to find 'kitti' dataset at '%s'" % path)
+
+        # TODO: should be removed when proper label merging is implemented
+        conflicting_types = {KittiTask.segmentation, KittiTask.detection}
```
Comment (zhiltsov-max) [inline datumaro/plugins/kitti_format/importer.py:36]: Why is it needed here? Can this format contain multiple unrelated datasets?

**Review by zhiltsov-max — CHANGES_REQUESTED** [review]
Please fix the remaining 2 questions.

File: datumaro/plugins/kitti_format/importer.py (hunk 15 above)
Comment (zmaslova) [reply datumaro/plugins/kitti_format/importer.py:36]: As you can see [segmentation](https://github.com/openvinotoolkit/datumaro/blob/mz/kitti/datumaro/plugins/kitti_format/format.py#L21) and [detection](https://github.com/openvinotoolkit/datumaro/blob/mz/kitti/datumaro/plugins/kitti_format/format.py#L59) labels are different (in original kitti dataset), so I think, we should avoid usage these both tasks at the same time

File: datumaro/plugins/kitti_format/converter.py (hunk 3 above)
Comment (zmaslova) [reply datumaro/plugins/kitti_format/converter.py:189]: No

File: datumaro/plugins/kitti_format/importer.py (hunk 15 above)
Comment (zhiltsov-max) [reply datumaro/plugins/kitti_format/importer.py:36]: Ok, got you.

**Review by zhiltsov-max — APPROVED** [review]

**Pull request merged**